kobo/apps/hook/tests/hook_test_case.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import json
+
+from django.conf import settings
+from django.core.urlresolvers import reverse
+import responses
+from rest_framework import status
+
+
+
+
+from ..constants import HOOK_LOG_FAILED
+from ..models import HookLog, Hook
+from kpi.constants import INSTANCE_FORMAT_TYPE_JSON
+from kpi.tests.kpi_test_case import KpiTestCase
+
+
+class HookTestCase(KpiTestCase):
+
+    def setUp(self):
+        self.client.login(username="someuser", password="someuser")
+        self.asset = self.create_asset(
+            "some_asset",
+            content=json.dumps({"survey": [{"type": "text", "name": "q1"}]}),
+            format="json")
+        self.asset.deploy(backend='mock', active=True)
+        self.asset.save()
+        self.hook = Hook()
+
+        v_uid = self.asset.latest_deployed_version.uid
+        submission = {
+            "__version__": v_uid,
+            "q1": u"¿Qué tal?",
+            "id": 1
+        }
+        self.asset.deployment._mock_submission(submission)
+        self.asset.save(create_version=False)
+        settings.CELERY_TASK_ALWAYS_EAGER = True
+
+    def _create_hook(self, return_response_only=False, **kwargs):
+        url = reverse("hook-list", kwargs={"parent_lookup_asset": self.asset.uid})
+        data = {
+            "name": kwargs.get("name", "some external service with token"),
+            "endpoint": kwargs.get("endpoint", "http://external.service.local/"),
+            "settings": kwargs.get("settings", {
+                "custom_headers": {
+                    "X-Token": "1234abcd"
+                }
+            })
+        }
+        response = self.client.post(url, data, format=INSTANCE_FORMAT_TYPE_JSON)
+        if return_response_only:
+            return response
+        else:
+            self.assertEqual(response.status_code, status.HTTP_201_CREATED,
+                             msg=response.data)
+            hook = self.asset.hooks.last()
+            self.assertTrue(hook.active)
+            return hook
+
+    def _send_and_fail(self):
+        """
+
+        The public method which calls this method, needs to be decorated by `@responses.activate
+        :return: dict
+        """
+        self.hook = self._create_hook()
+
+        ServiceDefinition = self.hook.get_service_definition()
+        submissions = self.asset.deployment.get_submissions()
+        instance_id = submissions[0].get("id")
+        service_definition = ServiceDefinition(self.hook, instance_id)
+        first_mock_response = {"error": "not found"}
+
+        # Mock first requests try
+        responses.add(responses.POST, self.hook.endpoint,
+                      json=first_mock_response, status=status.HTTP_404_NOT_FOUND)
+
+        # Mock next requests tries
```

Comment:
probably `requests' tries` (if it's plural). this is for subsequent mocked calls outside this `_send_and_fail()` method?

Reply:
Yep. It seems that requests can only be mocked this way if `responses.add` is called in the same scope. 
It would have been more obvious to me to move the "next tries" mock to the method that makes subsequent requests to `self.hook.endpoint` but it didn't work.